lib/AST/Util.cpp

```diff
@@ -25,22 +25,25 @@ unsigned GetHash(clang::ASTContext &ctx, clang::Stmt *stmt) {
 static bool IsEquivalent(clang::ASTContext &ctx, clang::Stmt *a, clang::Stmt *b,
                          llvm::FoldingSetNodeID &foldingSetA,
                          llvm::FoldingSetNodeID &foldingSetB) {
+  foldingSetA.clear();
+  foldingSetB.clear();
   a->Profile(foldingSetA, ctx, /*Canonical=*/true);
   b->Profile(foldingSetB, ctx, /*Canonical=*/true);
 
-  if (foldingSetA != foldingSetB) {
-    return false;
+  if (foldingSetA == foldingSetB) {
+    return true;
   }
 
-  CHECK_EQ(a->getStmtClass(), b->getStmtClass())
-      << "Statement classes differ but folding sets don't";
+  if (a->getStmtClass() != b->getStmtClass()) {
```

Comment:
I'd move this up first, before any folding set profiling, along with an identify check. I.e.

```c++
if (a == b) {
  return true;
} else if (a->getStmtClass() != b->getStmtClass()) {
  return false;
}

foldingSetA.clear();
...
```